File: urbanairship-automation/src/test/java/com/urbanairship/automation/actions/ScheduleActionTest.java
/* Copyright Airship and Contributors */

package com.urbanairship.automation.actions;

import com.urbanairship.PendingResult;
import com.urbanairship.actions.Action;
import com.urbanairship.actions.ActionArguments;
import com.urbanairship.actions.ActionResult;
import com.urbanairship.actions.ActionValue;
import com.urbanairship.automation.InAppAutomation;
import com.urbanairship.automation.Schedule;
import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonMap;
import com.urbanairship.json.JsonValue;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertTrue;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(AndroidJUnit4.class)
public class ScheduleActionTest {

    private InAppAutomation automation;
    private JsonMap scheduleJson;
    private ScheduleAction action;

    @Action.Situation
    private int[] acceptedSituations;

    @Action.Situation
    private int[] rejectedSituations;

    @Before
    public void setup() {
        this.automation = mock(InAppAutomation.class);

        this.action = new ScheduleAction(new Callable<InAppAutomation>() {
            @Override
            public InAppAutomation call() {
                return automation;
            }
        });

        // Accepted situations
        this.acceptedSituations = new int[] {
                Action.SITUATION_PUSH_RECEIVED,
                Action.SITUATION_MANUAL_INVOCATION,
                Action.SITUATION_WEB_VIEW_INVOCATION,
                Action.SITUATION_AUTOMATION
        };

        // Rejected situations
        this.rejectedSituations = new int[] {
                Action.SITUATION_PUSH_OPENED,
                Action.SITUATION_FOREGROUND_NOTIFICATION_ACTION_BUTTON,
                Action.SITUATION_BACKGROUND_NOTIFICATION_ACTION_BUTTON
        };

        List<JsonMap> triggersJson = new ArrayList<>();
        triggersJson.add(JsonMap.newBuilder()
                                .put("type", "foreground")
                                .put("goal", 20.0)
                                .build());

        scheduleJson = JsonMap.newBuilder()
                              .put("actions", JsonMap.newBuilder()
                                                     .put("tag_action", "cat")
                                                     .build())
                              .put("triggers", JsonValue.wrapOpt(triggersJson))
                              .put("group", "campaign")
                              .put("limit", 10)
                              .build();
    }

    /**
     * Test accepts arguments.
     */
    @Test
    public void testAcceptsArguments() {
        for (@Action.Situation int situation : acceptedSituations) {
            ActionArguments args = createArgs(situation, scheduleJson);
            assertTrue("Should accept arguments in situation " + situation,
                    action.acceptsArguments(args));
        }

        for (@Action.Situation int situation : rejectedSituations) {
            ActionArguments args = createArgs(situation, scheduleJson);
            assertFalse("Should reject arguments in situation " + situation,
                    action.acceptsArguments(args));
        }
    }

    private ActionArguments createArgs(int situation, JsonMap value) {
        return new ActionArguments(situation, ActionValue.wrap(value), null);
    }

    /**
     * Test perform with valid JSON.
     */
    @Test
    public void testPerform() {
        PendingResult<Boolean> pendingResult = new PendingResult<>();
        pendingResult.setResult(true);

        when(automation.schedule(any(Schedule.class))).thenReturn(pendingResult);

        ActionResult result = action.perform(createArgs(Action.SITUATION_MANUAL_INVOCATION, scheduleJson));
        assertNotNull(result.getValue().getString());
    }

    /**
     * Test perform with invalid JSON.
     */
    @Test
    public void testPerformInvalidJson() {
        ActionResult result = action.perform(createArgs(Action.SITUATION_MANUAL_INVOCATION, JsonMap.EMPTY_MAP));
        assertNotNull(result.getException());
        assertEquals(ActionResult.STATUS_EXECUTION_ERROR, result.getStatus());
    }

}

File: urbanairship-automation/src/test/java/com/urbanairship/automation/actions/ScheduleActionTest.kt
package com.urbanairship.automation.actions

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.urbanairship.actions.Action
import com.urbanairship.actions.ActionArguments
import com.urbanairship.actions.ActionValue
import com.urbanairship.automation.AutomationSchedule
import com.urbanairship.automation.EventAutomationTriggerType
import com.urbanairship.iam.actions.ScheduleAction
import com.urbanairship.json.JsonException
import com.urbanairship.json.jsonMapOf
import com.urbanairship.util.DateUtils
import junit.framework.TestCase.assertEquals
import junit.framework.TestCase.assertFalse
import junit.framework.TestCase.assertNull
import junit.framework.TestCase.assertTrue
import kotlinx.coroutines.test.TestResult
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertThrows
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
public class ScheduleActionTest {
    private var scheduled: AutomationSchedule? = null
    private val action = ScheduleAction(scheduler = { scheduled = it })

    @Test
    public fun testAcceptsArguments() {
        val valid = listOf(
            Action.SITUATION_MANUAL_INVOCATION,
            Action.SITUATION_WEB_VIEW_INVOCATION,
            Action.SITUATION_PUSH_RECEIVED,
            Action.SITUATION_AUTOMATION
        )

        val rejected = listOf(
            Action.SITUATION_FOREGROUND_NOTIFICATION_ACTION_BUTTON,
            Action.SITUATION_BACKGROUND_NOTIFICATION_ACTION_BUTTON,
            Action.SITUATION_PUSH_OPENED
        )

        for (situation in valid) {
            assertTrue(action.acceptsArguments(ActionArguments(situation, null, null)))
        }

        for (situation in rejected) {
            assertFalse(action.acceptsArguments(ActionArguments(situation, null, null)))
        }
    }

    @Test
    public fun testSchedule(): TestResult = runTest {
        val scheduleJson = jsonMapOf(
            "id" to "test-id",
            "type" to "actions",
            "group" to "test-group",
            "limit" to 1,
            "actions" to jsonMapOf(
                "action-name" to "action-value"
            ),
            "start" to DateUtils.createIso8601TimeStamp(0),
            "end" to DateUtils.createIso8601TimeStamp(1000),
            "triggers" to listOf(
                jsonMapOf(
                    "type" to "foreground",
                    "goal" to 2.0
                )
            ),
            "created" to DateUtils.createIso8601TimeStamp(0),
        )

        assertNull(scheduled)

        val scheduleID = action.perform(ActionArguments(Action.SITUATION_AUTOMATION, ActionValue.wrap(scheduleJson), null)).value.string
        assertEquals("test-id", scheduleID)

        assertEquals("test-id", scheduled?.identifier)
        assertEquals("test-group", scheduled?.group)
        assertEquals(1U, scheduled?.limit)
        assertEquals(1000UL, scheduled?.endDate)
        assertEquals(0UL, scheduled?.startDate)
        assertEquals(1, scheduled?.triggers?.size)
        assertEquals(EventAutomationTriggerType.FOREGROUND.value, scheduled?.triggers?.first()?.type)
        assertEquals(2.0, scheduled?.triggers?.first()?.goal)

        val actionJson = scheduled?.data as? AutomationSchedule.ScheduleData.Actions
        assertEquals(jsonMapOf(
            "action-name" to "action-value"
        ), actionJson?.actions)
    }

    @Test
    public fun testScheduleThrowsOnInvalidSource(): TestResult = runTest {
        assertThrows(JsonException::class.java) { action.perform(ActionArguments(
            Action.SITUATION_AUTOMATION,
            ActionValue.wrap(jsonMapOf()),
            null))}
    }
}
